### PR TITLE
Make color editable

### LINKED ISSUE #6: Make color editable
To allow merchants to apply their style guide to the plugin, the merchants should be able to adjust the colors. 

## Used colors 
- ~Font color~
- ~Background color of outer progress bar~
- ~Background color of inner progress bar~
- ~Border color of outer progress bar~
- Message color
- Progress bar color

### PR DESCRIPTION
**Pull request by nielslange**
Fixes #6 

## Note

In i1 of this block, the merchant must be able to select the color of the message and the progress bar. This PR aims to add this functionality.

### Named colors

<table>
<tr>
<td valign="top">Editor:
<br><br>

#14-editor-named-colors
</td>
<td valign="top">Frontend:
<br><br>

#14-frontend-named-colors
</td>
</tr>
</table>

### Custom colors

<table>
<tr>
<td valign="top">Editor:
<br><br>

#14-editor-custom-colors
</td>
<td valign="top">Frontend:
<br><br>

#14-frontend-custom-colors
</td>
</tr>
</table>

## Test instructions

* Check out this PR.
* Create a test page.
* Add the block Free Shipping Progress Bar to the test page.
* Click on the block to open the Inspector Controls.
* Adjust the message color and the progress bar color using named colors.
* Ensure that the color change in the editor.
* Save the test page.
* Ensure that the color change in the frontend.
* Adjust the message color and the progress bar color using custom colors.
* Ensure that the color change in the editor.
* Save the test page.
* Ensure that the color change in the frontend.

### CODE REVIEW

File: src/edit.js [hunk 1]
```diff
@@ -42,6 +46,34 @@ export default function Edit( { attributes, setAttributes } ) {
 							} )
 						}
 					/>
+					<PanelColorSettings
```
Comment (senadir) [inline src/edit.js]: Thank you for working on this @nielslange 
Unfortunately, this is not how we want to use colors.
Your implementation only implements custom colors (as in, the color will always be a hex/rgb code).

In Gutenberg, you have the option between a relative color and a custom color. Relative colors are defined by a theme in a palette, and custom colors are custom colors.
Relative colors are applied by a className, custom colors are hardcoded directly.

I'd invite you to take a look into Gutenberg codebase, `packages/block-library/src` has all of Gutenberg blocks.
You can also check some of our atomic blocks like `js/atomic/block/product-elements/title`

You can also read this article I wrote a year ago about how to implement color controls in a block.
pca54o-6J-p2

**Review by senadir — CHANGES_REQUESTED** [review]

File: src/edit.js (hunk 1 above)
Comment (nielslange) [reply src/edit.js]: Thanks for your review and the suggestions, @senadir. I'll take a look at the resources you pointed out, to update this PR accordingly.

Comment (nielslange) [reply src/edit.js:64]: @senadir: As requested, I've implemented the colors using `withColors` and `getColorClassName`. Implementing the font color and the background color was no problem. However, to implement the border color, I had to copy the following CSS definitions from Gutenberg into this PR:

```css
.has-pale-pink-border-color {
	border-color: #f78da7;
}
.has-vivid-red-border-color {
	border-color: #cf2e2e;
}
.has-luminous-vivid-orange-border-color {
	border-color: #ff6900;
}
.has-luminous-vivid-amber-border-color {
	border-color: #fcb900;
}
.has-light-green-cyan-border-color {
	border-color: #7bdcb5;
}
.has-vivid-green-cyan-border-color {
	border-color: #00d084;
}
.has-pale-cyan-blue-border-color {
	border-color: #8ed1fc;
}
.has-vivid-cyan-blue-border-color {
	border-color: #0693e3;
}
.has-vivid-purple-border-color {
	border-color: #9b51e0;
}
.has-white-border-color {
	border-color: #fff;
}
.has-very-light-gray-border-color {
	border-color: #eee;
}
.has-cyan-bluish-gray-border-color {
	border-color: #abb8c3;
}
.has-very-dark-gray-border-color {
	border-color: #313131;
}
.has-black-border-color {
	border-color: #000;
}
```

I saw that you created https://github.com/WordPress/gutenberg/issues/16848, but this hasn't been addressed yet. I wonder if my approach is the correct way or if you can think of a better way.

**Comment by nielslange** [comment]
@senadir: As requested, I've implemented the colors using `withColors` and `getColorClassName`. Implementing the font color and the background color was no problem. However, to implement the border color, I had to copy the following CSS definitions from Gutenberg into this PR:

```css
.has-pale-pink-border-color {
	border-color: #f78da7;
}
.has-vivid-red-border-color {
	border-color: #cf2e2e;
}
.has-luminous-vivid-orange-border-color {
	border-color: #ff6900;
}
.has-luminous-vivid-amber-border-color {
	border-color: #fcb900;
}
.has-light-green-cyan-border-color {
	border-color: #7bdcb5;
}
.has-vivid-green-cyan-border-color {
	border-color: #00d084;
}
.has-pale-cyan-blue-border-color {
	border-color: #8ed1fc;
}
.has-vivid-cyan-blue-border-color {
	border-color: #0693e3;
}
.has-vivid-purple-border-color {
	border-color: #9b51e0;
}
.has-white-border-color {
	border-color: #fff;
}
.has-very-light-gray-border-color {
	border-color: #eee;
}
.has-cyan-bluish-gray-border-color {
	border-color: #abb8c3;
}
.has-very-dark-gray-border-color {
	border-color: #313131;
}
.has-black-border-color {
	border-color: #000;
}
```

I saw that you created https://github.com/WordPress/gutenberg/issues/16848, but this hasn't been addressed yet. I wonder if my approach is the correct way or if you can think of a better way.

File: src/index.js [hunk 2]
```diff
@@ -34,7 +35,10 @@ registerBlockType( 'nielslange/free-shipping-progress-bar', {
 	/**
 	 * @see ./edit.js
 	 */
-	edit: Edit,
+	// edit: Edit,
+	edit: withColors( { messageColor: 'color', progressColor: 'color' } )(
+		Edit
+	),
```
Comment (senadir) [inline src/index.js]: Refrain from calling a HOC here. All Edit functionality should be exported as part of Edit component, so this HOC should be there.

File: src/block.js [hunk 3]
```diff
@@ -1,12 +1,42 @@
 /**
  * External dependencies
  */
+import { getColorClassName } from '@wordpress/block-editor';
```
Comment (senadir) [inline src/block.js]: This is out of the scope of your level, but `@wordpress/block-editor` shouldn't be called on the frontend. `block-editor` is a heavy package, that would severely hinder performance on the frontend.

Given that you're only using `getColorClassName`, can you just copy the function in an util file. Make sure to import `kebabCase` like this.

```
npm install lodash.kebabcase
```

```js
import kebabCase from 'lodash.kebabcase'

export function getColorClassName( colorContextName, colorSlug ) {
	if ( ! colorContextName || ! colorSlug ) {
		return undefined;
	}

	return `has-${ kebabCase( colorSlug ) }-${ colorContextName }`;
}
```

**Review by senadir — COMMENTED** [review]
I took a quick glance, I will be waiting for when you update this to include `currenColor`

File: src/index.js (hunk 2 above)
Comment (nielslange) [reply src/index.js]: Thanks for your feedback. As requested, I moved the `withColors()` HOC from `index.js` to `edit.js`.

File: src/block.js (hunk 3 above)
Comment (nielslange) [reply src/block.js]: Thanks for pointing this out. I just adjusted the code accordingly and updated this PR.

File: block.json [hunk 4]
```diff
@@ -17,6 +17,12 @@
 		"freeShippingFrom": {
 			"type": "number",
 			"default": "50"
+		},
+		"messageColor": {
+			"type": "string"
+		},
+		"progressColor": {
+			"type": "string"
```
Comment (senadir) [inline block.json]: This alone isn't enough. You're not just saving one form of colors (which is just the name here). Your attributes should be two fold: color, and customColor. So, `messageColor`, `customMessageColor`. This is what we did in our blocks.

File: src/block.js [hunk 5]
```diff
@@ -1,12 +1,39 @@
 /**
  * External dependencies
  */
+import { getColorClassName } from './util';
 import { __, sprintf } from '@wordpress/i18n';
+import classnames from 'classnames';
+
+export default function Block( {
+	color,
```
Comment (senadir) [inline src/block.js]: What's color here?

File: src/block.js [hunk 6]
```diff
@@ -24,9 +51,22 @@ export default function Block( { currentTotal, freeShippingFrom } ) {
...
+					progressBorderClasses
+				) }
+			>
+				<div
+					className={ classnames(
+						'progress-bar',
+						progressBackgroundClasses
+					) }
+					style={ progressBarStyle }
+				></div>
```
Comment (senadir) [inline src/block.js]: Your changes should be immediately reflected on the editor, they're not.
image

File: src/edit.js [hunk 7]
```diff
@@ -25,26 +31,70 @@ import Block from './block';
...
+							'Provide the value in your store currency.',
+							'free-shipping-progress-bar'
+						) }
 						value={ freeShippingFrom }
 						onChange={ ( value ) =>
 							setAttributes( {
 								freeShippingFrom: value,
 							} )
 						}
 					/>
```
Comment (senadir) [inline src/edit.js]: Outside scope of this PR, but I just saw this. This value is not persisting, but I won't tell you yet why 😉

File: src/edit.js [hunk 8]
```diff
@@ -25,26 +31,70 @@ import Block from './block';
...
+							{
+								label: __(
+									'Progress Bar Color',
+									'free-shipping-progress-bar'
+								),
+								value: progressColor.color,
+								onChange: setProgressColor,
+							},
+						] }
+					></PanelColorSettings>
```
Comment (senadir) [inline src/edit.js]: This is correct, however, you're not consuming the values correctly.

File: src/progress-bar-icon.js [hunk 9]
```diff
@@ -48,16 +48,16 @@ export default function ProgressBarIcon() {
 							xlinkHref="#a3HIePHoOc"
```
Comment (senadir) [inline src/progress-bar-icon.js]: Changes in this file are outside the scope of this PR, why are they here?

File: src/style.scss [hunk 10]
```diff
@@ -1,21 +1,18 @@
...
+	border-radius: 0.125em;
+	display: flex;
+	height: 2em;
+	justify-content: center;
+	margin-top: 0.5em;
+	position: relative;
 }
+
+.progress-bar {
+	background: #96588a;
```
Comment (senadir) [inline src/style.scss]: You're still providing a default color here, when you shouldn't.

File: src/util.js [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import kebabCase from 'lodash.kebabcase';
```
Comment (senadir) [inline src/util.js:4]: 👍

**Review by senadir — CHANGES_REQUESTED** [review]
Hey Niels, this is not working so far. You're on the correct track, but not there yet.

Please make sure to test your PR before making it available for a review. This might have been working on a previous iteration but isn't working right now and should have been fully retested.

If it's actually working for you fine, then we can try to understand why my version isn't working yet.

File: block.json (hunk 4 above)
Comment (nielslange) [reply block.json]: I removed `customMessageColor` and `customProgressColor` as the color settings are working fine on my end without these definitions. I'll add them again, but it would be nice to take a look at my local site to find out why the code works as expected on my end.

File: src/block.js (hunk 5 above)
Comment (nielslange) [reply src/block.js]: I must have missed this when cleaning up unused variables. I just removed it.

File: src/block.js (hunk 6 above)
Comment (nielslange) [reply src/block.js]: On my end, the changes take place immediately:

Screen Capture on 2021-09-30 at 22-39-10

File: src/edit.js (hunk 7 above)
Comment (nielslange) [reply src/edit.js]: Oh, it seems I have defined the wrong attribute type. I changed the type from `number` to `string`. Now the variable gets stored as expected.

File: src/edit.js (hunk 8 above)
Comment (nielslange) [reply src/edit.js]: I do not see what's wrong here. Isn't this identical to https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/e23b51119dfa5c11e67ab0b3cbd89d49189a8569/assets/js/atomic/blocks/product-elements/title/edit.js#L90-L105?

File: src/progress-bar-icon.js (hunk 9 above)
Comment (nielslange) [reply src/progress-bar-icon.js]: I applied these changes as I noticed that they're causing an error in the console. To avoid this error from happening, I applied this fix. Shall I revert it and create a separate PR for it?

File: src/style.scss (hunk 10 above)
Comment (nielslange) [reply src/style.scss]: I thought that's what I need to do to get the default background color to work when someone clicks the `clear` button.  When removing this color, the progress bar is not visible when being added to the page and it becomes invisible when the `clear` button gets clicked. 

Screenshot 2021-09-30 at 23 12 27

How can I remove the color code while ensuring that the progress bar has a default color when getting added and when the `clear` button gets clicked?

**Comment by nielslange** [comment]
> Please make sure to test your PR before making it available for a review. This might have been working on a previous iteration but isn't working right now and should have been fully retested.

It's strange that the PR was is working on your end, as it works as expected on my end. Before pushing a change, I always check if the change is working as expected. Of course, I might oversee one or the other issue now and then. But in general, I do not push untested code and requesting a review. 

> If it's actually working for you fine, then we can try to understand why my version isn't working yet.

It would be great if we could look at my code together, to see why it's working on my end, to prevent that this inconsistency is happening again in the future.

File: src/edit.js (hunk 7 above)
Comment (senadir) [reply src/edit.js]: The value should be however a number to do arithmetics on it.

File: src/edit.js (hunk 8 above)
Comment (senadir) [reply src/edit.js]: Yes, this code part is correct, I was asserting that no changes are needed here, you just need to change how you consume the colors above.

File: src/block.js (hunk 6 above)
Comment (senadir) [reply src/block.js]: I figured out why they're not running for me. I'm running Gutenberg plugin.

File: src/progress-bar-icon.js (hunk 9 above)
Comment (senadir) [reply src/progress-bar-icon.js]: It's okay, however, in the future, keep those changes separate.

File: src/style.scss (hunk 10 above)
Comment (senadir) [reply src/style.scss]: How about using `currentColor` there, and it would get the text color?

**Comment by senadir** [comment]
> It would be great if we could look at my code together, to see why it's working on my end, to prevent that this inconsistency is happening again in the future.

This was a dependency issue, as I was running Gutenberg plugin. We support Gutenberg in our code so we should try to fix this if it's possible.

**Comment by senadir** [comment]
Also, custom colors are not working, Gutenberg active or not, this is because they're not being used in your code, and you're only using the color tokens.

File: src/style.scss (hunk 10 above)
Comment (nielslange) [reply src/style.scss]: > How about using currentColor there, and it would get the text color?
Thanks for pointing this out. I haven't thought about using `currentColor` in such a way. I just adjusted it.

File: src/progress-bar-icon.js (hunk 9 above)
Comment (nielslange) [reply src/progress-bar-icon.js]: I removed  that part from this PR and created a separate PR for it.

File: src/edit.js (hunk 8 above)
Comment (nielslange) [reply src/edit.js]: I'm not sure which line you're referring to. Would you mind telling me the line number, so that can see what you mean?

**Comment by nielslange** [comment]
> Also, custom colors are not working, Gutenberg active or not, this is because they're not being used in your code, and you're only using the color tokens.

I replaced the current color settings with Gutenberg's native color support. Both color variants and custom colors are now working both in the editor and on the frontend independent of the activated theme.

You can go ahead and do another review, once you have time doing so. 😀

**Comment by senadir** [comment]
While both custom and predefined colors are working fine. I'm seeing a single option now. Is this the expected behavior now? if so, can you update the PR description and double check in the project thread?
image

File: src/edit.js [hunk 12]
```diff
@@ -42,3 +50,10 @@ export default function Edit( { attributes, setAttributes } ) {
 		</div>
 	);
 }
+
+const Edit = withColors( 'color', {
+	messageColor: 'color',
+	progressColor: 'color',
+} )( ProgressEdit );
+
```
Comment (senadir) [inline src/edit.js]: If you switched to a pure Gutenberg implementation, this shouldn't be needed anymore.

However, I don't think we fully switched yet, as we still need both anyway, but this isn't consumed above, so it's just sitting here adding attributes for nothing.

**Review by senadir — CHANGES_REQUESTED** [review]

File: src/edit.js (hunk 12 above)
Comment (nielslange) [reply src/edit.js]: I decided to switch to a pure Gutenberg implementation as it works in all themes, while the previous implementation did not work with all themes, such as Storefront.

**Comment by nielslange** [comment]
> While both custom and predefined colors are working fine. I'm seeing a single option now. Is this the expected behavior now? if so, can you update the PR description and double check in the project thread?

I've updated the screenshots in the PR description accordingly and in the next weekly update in the project thread, I'll point out why I moved to the native Gutenberg color settings and why I'm only using the text color settings and not the background color settings.

File: src/util.js [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+ */
+import kebabCase from 'lodash.kebabcase';
+
+export function getColorClassName( colorContextName, colorSlug ) {
+	if ( ! colorContextName || ! colorSlug ) {
+		return undefined;
+	}
+
+	return `has-${ kebabCase( colorSlug ) }-${ colorContextName }`;
+}
```
Comment (senadir) [inline src/util.js]: This file as a whole isn't used.

**Review by senadir — CHANGES_REQUESTED** [review]
I see you removed the background color option. I think we can use a mix of `withColor` and `supports.color` to achieve both goals.
It's true that `supports.color` is the recommended way to add colors, but it doesn't fit all use cases, and in that case, we can also `withColor`

File: src/util.js (hunk 13 above)
Comment (nielslange) [reply src/util.js]: If we move back to a mix of `withColor` and `supports.color`, don't we need this file again then?

Comment (senadir) [reply src/util.js]: Probably yeah.

**Comment by nielslange** [comment]
@senadir I adjusted the color settings using the native Gutenberg color support for both the text and the background color. Initially, I faced the problem that the custom colors were only visible in the editor. I overcame this with the following section:

https://github.com/woocommerce/woocommerce-free-shipping-progress-bar-block/blob/fc403cd0bb29f3dadea6edbd2d5e77c9a32424a5/woocommerce-free-shipping-progress-bar-block.php#L47-L55

This section adds the custom colors to the attributes so that they can be added to the block as data-attributes. Before this change, the outer element of my block looked like this:

```html
<div data-style="{"color":{"background":"#bf616a","text":"#5e81ac"}}" data-current-total="15" data-free-shipping-from="50" class="wp-block-nielslange-free-shipping-progress-bar">....</div>
```

Due to my change, the outer element now looks like this:

```html
<div data-background-color="#bf616a" data-text-color="#5e81ac" data-current-total="15" data-free-shipping-from="50" class="wp-block-nielslange-free-shipping-progress-bar">...</div>
```

In addition, I noticed that not all themes were showing the named colors in the editor, for example, the [Twenty Nineteen theme](https://wordpress.org/themes/twentynineteen/). I overcame this problem with the following section:

https://github.com/woocommerce/woocommerce-free-shipping-progress-bar-block/blob/fc403cd0bb29f3dadea6edbd2d5e77c9a32424a5/src/block.js#L22-L43

I'm excited to hear your thoughts on these changes.

File: src/block.js [hunk 14]
```diff
@@ -2,12 +2,50 @@
...
+	};
+
+	// Prepare progress class and style.
+	const progressColor =
+		backgroundColor ?? colorProps?.style?.backgroundColor ?? 'currentColor';
+	const progressClass = getColorClassName( 'color', progressColor );
+	const progressStyle = {
+		width: progressWidth,
+		color: colorProps?.style?.backgroundColor ?? progressColor,
+	};
```
Comment (senadir) [inline src/block.js]: Given that `Block` would never get `colorProps` and you're doing a lot of massaging to select either colorProps or attributes, why not move this logic to `Edit` and do the massaging there, ideally, `Block` wouldn't care from where the colors and classes are coming from, it should just apply style/class.

File: src/block.js [hunk 15]
```diff
@@ -2,12 +2,50 @@
...
+
+	// Calculate remaining value.
 	const remaining = Number( freeShippingFrom - currentTotal ).toFixed( 2 );
+
+	// Prepare message. v
```
Comment (senadir) [inline src/block.js]:
```suggestion
	// Prepare message.
```

File: woocommerce-free-shipping-progress-bar-block.php [hunk 16]
```diff
@@ -44,13 +44,29 @@ function enqueue_frontend_script() {
...
+	// Add the custom background color to the attributes array.
+	if ( ! $attributes['backgroundColor'] && $attributes['style']['color']['background'] ) {
+		$attributes['backgroundColor'] = $attributes['style']['color']['background'];
+	}
+
+	// Add the custom text color to the attributes array.
+	if ( ! $attributes['textColor'] && $attributes['style']['color']['text'] ) {
+		$attributes['textColor'] = $attributes['style']['color']['text'];
+	}
+
```
Comment (senadir) [inline woocommerce-free-shipping-progress-bar-block.php]: I'd like to understand what's happening here really? this logic seems redundant.

Comment (nielslange) [reply woocommerce-free-shipping-progress-bar-block.php]: Thanks for asking. In https://github.com/woocommerce/woocommerce-free-shipping-progress-bar-block/pull/14#issuecomment-965168997, I added a bit more context on that, but in a nutshell, when saving custom colors, the div on the frontend looks like this:

```
<div data-style="{"color":{"background":"#bf616a","text":"#5e81ac"}}" data-current-total="15" data-free-shipping-from="50" class="wp-block-nielslange-free-shipping-progress-bar">....</div>
```

The code, you pointed out, converts this to:

```
<div data-background-color="#bf616a" data-text-color="#5e81ac" data-current-total="15" data-free-shipping-from="50" class="wp-block-nielslange-free-shipping-progress-bar">...</div>
```

That way, the custom colors can be picked up correctly when rendering this component. It definitely looks redundant. It is a slightly simplified version of https://github.com/WordPress/gutenberg/blob/2d2e51ee954bf26f3e1a1a4187d0913a31f9459d/packages/block-library/src/search/index.php#L311-L352 though, and I cannot think of a more efficient way of adding custom colors as data-attributes. 

Do you have a suggestion on how to simplify this part?

Comment (senadir) [reply woocommerce-free-shipping-progress-bar-block.php]: > Thanks for asking. In #14 (comment), I added a bit more context on that, but in a nutshell,

Ah yeah, I missed that, thanks for pointing out.

> That way, the custom colors can be picked up correctly when rendering this component. It definitely looks redundant. It is a slightly simplified version of

Handling colors is still a bit tricky right now in Gutenberg, I managed to get this to work another way, using `useBorderProps` but it's experimental. I think this solution is by itself good and does the job, we can probably see where things are in GB later down the road.

File: src/block.js (hunk 14 above)
Comment (nielslange) [reply src/block.js]: > Given that Block would never get colorProps and you're doing a lot of massaging to select either colorProps or attributes, why not move this logic to Edit and do the massaging there, ideally, Block wouldn't care from where the colors and classes are coming from, it should just apply style/class.

I looked into this, but as this block is a dynamic block, the colors for the frontend need to be calculated on the fly. When I move the color logic into `Edit`, the frontend no longer receives any colors, unless I'm missing something. 

Do you want to run a quick test to see if you're able to move the color logic into `Edit` while ensuring that the colors remain visible in the frontend?

File: woocommerce-free-shipping-progress-bar-block.php (hunk 16 above)
Comment (nielslange) [reply woocommerce-free-shipping-progress-bar-block.php]: >  I think this solution is by itself good and does the job, we can probably see where things are in GB later down the road.

That sounds like a good plan.

File: src/block.js (hunk 14 above)
Comment (senadir) [reply src/block.js]: > I looked into this, but as this block is a dynamic block, the colors for the frontend need to be calculated on the fly.

I don't see how would this makes a different? the inputs we use to calculate colors are just attributes, which are also available on the editor? The only different between editor and frontend is cart data, which is not included.

I can give this a try later this week, but I think you should be able to do it with some code massage, unless I'm missing another input that's only available in frontend?

Comment (nielslange) [reply src/block.js]: Thanks for your feedback, Nadir. I'll take another look to see how I can avoid the color calculation within the block and move that logic completely to `Edit`, so that `Block` only renders the colors.

File: woocommerce-free-shipping-progress-bar-block.php [hunk 17]
```diff
@@ -44,20 +44,37 @@ function enqueue_frontend_script() {
 function add_attributes_to_block( $attributes = [], $content = '' ) {
 	$escaped_data_attributes = [];
 
+	// Add the custom background color to the attributes array.
+	if ( ! $attributes['backgroundColor'] && $attributes['style']['color']['background'] ) {
```
Comment (mikejolley) [inline woocommerce-free-shipping-progress-bar-block.php]: I got a bunch of warnings from these lines. Like JS, if the key is undefined and you look deeper, you'll get notices in PHP.

You could use an `isset`:

```
if ( isset( $attributes['style'], $attributes['style']['color'], $attributes['style']['color']['background'] ) ) {
```

File: block.json [hunk 18]
```diff
@@ -7,7 +7,10 @@
 	"category": "woocommerce",
 	"parent": [ "woocommerce/cart-totals-block" ],
 	"supports": {
-		"html": false
+		"html": false,
+		"color": {
+			"__experimentalSkipSerialization": true
```
Comment (mikejolley) [inline block.json:12]: Can you tell me what this does please?

File: src/util.js [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+	) }-${ colorContextName } has-text-color`;
+}
+
+export function getColorCode( colorCode, colorProps, scope ) {
+	if ( colorProps !== undefined && colorProps.style !== undefined ) {
+		return colorProps.style[ scope ];
+	}
+
+	if ( colorCode !== undefined ) {
+		return colorCode;
```
Comment (mikejolley) [inline src/util.js]: Personally prefer shorthand for these, but up to you:

```
return colorCode || 'currentColor';
```

**Review by mikejolley — CHANGES_REQUESTED** [review]
Left some comments. The biggest issue are the PHP notices on first running this PR. They need to be fixed with `isset`.

R/e the functionality, I think the main thing lacking (from experience using similar customisation controls on other platforms) is control of both the progress indicator, and the track. So if possible, can you add separate controls for those colours?

As well as progress and track, the border is always grey. What about generating a border colour from the track colour? Or just using track for border too.

Finally, and not related entirely to this PR, have you considered using the HTML5 `<progress>` element instead of a `div`?

File: block.json (hunk 18 above)
Comment (nielslange) [reply block.json:12]: This part ensures that the background color setting does not affect the background color of the entire block. On https://github.com/WordPress/gutenberg/issues/32417 and https://github.com/WordPress/gutenberg/issues/34216#issuecomment-904950890 you can find further information on this.

File: woocommerce-free-shipping-progress-bar-block.php (hunk 17 above)
Comment (nielslange) [reply woocommerce-free-shipping-progress-bar-block.php]: Thanks for pointing this out. I totally missed that, and I'll add some tools, e.g. phpcs, Psalm, etc. to the plugin so that I can spot mistakes like these easier.

File: src/util.js (hunk 19 above)
Comment (nielslange) [reply src/util.js]: Your suggestion looks much better! I also prefer shorthand syntax. Suggestion applied. 🙌

**Comment by nielslange** [comment]
> Left some comments. The biggest issue are the PHP notices on first running this PR. They need to be fixed with isset.

Great catch and I've already adjusted that.

> R/e the functionality, I think the main thing lacking (from experience using similar customisation controls on other platforms) is control of both the progress indicator, and the track. So if possible, can you add separate controls for those colours?

> As well as progress and track, the border is always grey. What about generating a border colour from the track colour? Or just using track for border too.

Funny enough, in an earlier version, the border color was derived from the process bar background color. I then moved away from that solution towards using the same color as for the quantity buttons. That said, I moved back to the initial version of using the same color both for the process bar and the border.

> Finally, and not related entirely to this PR, have you considered using the HTML5 `<progress>` element instead of a `div`?

I have, and I voted against that for design reasons. The HTML5 `<progress>` element offers very limited styling options and even these already limited options are not supported by every browser. To give merchants the highest level of flexibility, I decided to implement the progress bar using an outer and an inner div. That way, it's possible to style the element cross-browser compatible.

Since I applied all changes, would you mind doing another review, @mikejolley?

**Comment by mikejolley** [comment]
> The HTML5 <progress> element offers very limited styling options and even these already limited options are not supported by every browser.

Is this still true? I've used them before and I know css offers a reset so you essentially get 2 bars (track and progress): https://css-tricks.com/html5-progress-element/ I'd argue we should use semantic elements whenever possible even if it compromises slightly on appearance, because div soup has no meaning.

**Comment by nielslange** [comment]
> Is this still true? I've used them before and I know css offers a reset so you essentially get 2 bars (track and progress): https://css-tricks.com/html5-progress-element/ I'd argue we should use semantic elements whenever possible even if it compromises slightly on appearance, because div soup has no meaning.

I agree that we should prefer semantic elements.  I might have faced the problem before, since I haven't reset the default behavior. I'll give it another try.

**Comment by nielslange** [comment]
I switched to the native HTML5 progress element, and resetting the styles indeed solved the cross-browser problem I faced before. I checked the element in Chrome, Firefox, Opera and Safari and the styling looks identical in all browsers.

**Comment by mikejolley** [comment]
@nielslange bar looks good. I notice the currency symbol is missing in the message. Is this is separate issue or new?

**Comment by nielslange** [comment]
> @nielslange bar looks good. I notice the currency symbol is missing in the message. Is this is separate issue or new?

The currency symbol will be implemented with #29. Can I go ahead and merge this PR or did you notice anything else that should be adjusted?

**Review by mikejolley — APPROVED** [review]
